apio-consumer/src/main/kotlin/com/liferay/apio/consumer/ApioConsumer.kt

```diff
@@ -15,416 +15,95 @@
-		.entries
-		.fold(mutableMapOf<String, Any>() to mutableMapOf(), foldEntry(context))
-}
-
-typealias Embedded<T, U> = Map<T, U>
-typealias EmbeddedList<T> = List<T>
-typealias FoldedAttributes = Pair<MutableMap<String, Any>, MutableMap<String, Thing?>>
-
-private fun foldEntry(context: Context?) = { acc: FoldedAttributes, entry: Entry<String, Any> ->
-	val (attributes, things) = acc
-
-	val key = entry.key
-	val value = entry.value
-
-	when {
-		value is Embedded<*, *> -> value.parse(context, things, attributes, key)
-		value is EmbeddedList<*> && value.hasMapInstances() ->
-			value.parse(context, things, attributes, key)
-		context != null && context.isId(key) -> with(value as String) {
-			things[this] = null
-			attributes[key] = Relation(this)
-		}
-		else -> attributes[key] = value
-	}
-
-	attributes to things
-}
-
-private fun Embedded<*, *>.parse(context: Context?, things: MutableMap<String, Thing?>,
-								 attributes: MutableMap<String, Any>, key: String) {
-
-	(this as? Embedded<String, Any>)?.apply {
-		if (this.containsKey("@id")) {
-			val result = flatten(this, context)
-
-			result?.let {
-				val (thing, embeddedThings) = it
-				attributes[key] = Relation(thing.id)
-
-				things.put(thing.id, thing)
-
-				things.putAll(embeddedThings)
-			}
-		} else {
-			val (attr, embeddedThings) = foldTree(this, context)
-
-			things.putAll(embeddedThings)
-			attributes[key] = attr
-		}
-	}
-}
-
-private fun EmbeddedList<*>.parse(context: Context?, things: MutableMap<String, Thing?>,
-								  attributes: MutableMap<String, Any>, key: String) {
-
-	(this as? EmbeddedList<Embedded<String, Any>>)?.apply {
-		if (this.first().containsKey("@id")) {
-			val list = this.map { flatten(it, context)!! }
-			val mutableList = mutableListOf<Relation>()
-
-			for ((thing, embeddedThings) in list) {
-				mutableList.add(Relation(thing.id))
-
-				things.put(thing.id, thing)
-
-				things.putAll(embeddedThings)
-			}
-
-			attributes[key] = mutableList
-		} else {
-			val pairsList = this.map { foldTree(it, context) }
-
-			val foldedList = mutableListOf<Map<String, Any>>()
-
-			for ((list, embeddedThings) in pairsList) {
-				foldedList.add(list)
-
-				things.putAll(embeddedThings)
-			}
-
-			attributes[key] = foldedList
-		}
-
-	}
+/**
+ * @author Javier Gamarra
+ * @author Paulo Cruz
+ */
+object ApioConsumer {
+
+    fun fetch(url: HttpUrl, onSuccess: (Thing) -> Unit, onError: (Exception) -> Unit = emptyOnError()) {
+        fetch(url, emptyMap(), emptyList(), onSuccess, onError)
+    }
+
+    fun fetch(url: HttpUrl, fields: Map<String, List<String>> = emptyMap(), embedded: List<String> = emptyList(),
+        onSuccess: (Thing) -> Unit, onError: (Exception) -> Unit = emptyOnError()) {
+
+        fetch(url, fields, embedded) {
+            it.fold(onSuccess, onError)
+        }
+    }
+
+    @JvmOverloads
+    fun fetch(url: HttpUrl, fields: Map<String, List<String>> = emptyMap(), embedded: List<String> = emptyList(),
```

Comment:
Same here, why not having only one method with default values and @JvmOverloads for java support

Reply:
At least for me, using only JvmOverloads don't work very well in the Kotlin side, I needed to name the onSuccess and onError params since it wasn't inferred when I didn't pass second and third params. That's why I did another function. 

The other question is that now we have two different signatures, one with onComplete and other with onSuccess and onError callbacks. We did that because it's not so cool to use the Result in Java, so it's just a wrapping to the onComplete function.

@victorg1991 do you have a suggestion?

Reply:
I see, but you can still remove one method right? and have one method with onSucces/onError and one with onComplete

Reply:
Ok, I'll keep the method with onComplete.